contracts/presale/Presale.sol

```diff
@@ -48,15 +50,16 @@ contract Presale is Ownable {
 
     function buy(uint _amountUSDC) external {
         require(block.number <= startBlock + duration, "Presale: INVALID_DATE");
-        require(_amountUSDC > 0, "Presale: ZERO_AMOUNT_USDC");
         uint availableAmountSHBT = purchasedLimit.sub(purchasedAmountOf[msg.sender]);
         uint amountSHBT = _amountUSDC.mul(PresaleConstants.ONE_HBT_IN_WEI).div(rate);
         if (amountSHBT > availableAmountSHBT) {
-            _amountUSDC = amountSHBT.sub(availableAmountSHBT).mul(rate);
             amountSHBT = availableAmountSHBT;
+            _amountUSDC = amountSHBT.mul(rate).div(PresaleConstants.ONE_HBT_IN_WEI);
```

Comment:
why calculation of `_amountUSDC` should be changed? in this case we try to charge USDC amount for all available tokens regarding limits, instead of revert it

Reply:
**I think there is a calculation error here**.
Let's figure it out.

_Let the Presale contract have the following constant values_:
- rate = 15 * ONE_HBT_IN_WEI / 100
- purchaseLimit = 100000 * ONE_HBT_IN_WEI

_We have two options_: **B1** - before the changes, **B2** - after the changes.
- **B1**:
        uint availableAmountSHBT = purchasedLimit.sub(purchasedAmountOf[msg.sender]);
        uint amountSHBT = _amountUSDC.mul(PresaleConstants.ONE_HBT_IN_WEI).div(rate);
        if (amountSHBT > availableAmountSHBT) {
            _amountUSDC = amountSHBT.sub(availableAmountSHBT).mul(rate);
            amountSHBT = availableAmountSHBT;
        }
- **B2**:
        uint availableAmountSHBT = purchasedLimit.sub(purchasedAmountOf[msg.sender]);
        uint amountSHBT = _amountUSDC.mul(PresaleConstants.ONE_HBT_IN_WEI).div(rate);
        if (amountSHBT > availableAmountSHBT) {
            amountSHBT = availableAmountSHBT;
            _amountUSDC = amountSHBT.mul(rate).div(PresaleConstants.ONE_HBT_IN_WEI);
        }

For each of the options, _we will calculate two independent purchases_:
1. Purchase in the amount of _amountUSDC = 7500 (**purchaseLimit will not be exceeded**).
**B1** and **B2**: _amountUSDC = 7500, amountSHBT = 7500 * ONE_HBT_IN_WEI / (15 * ONE_HBT_IN_WEI / 100) = 50000. Since amountSHBT < purchaseLimit - no recalculation is needed.

2. Purchase in the amount of _amountUSDC = 75000 (**purchaseLimit will be exceeded**).
**В1**: _amountUSDC = 75000, amountSHBT = 75000 * ONE_HBT_IN_WEI / (15 * ONE_HBT_IN_WEI / 100) = 500000. Since amountSHBT> purchaseLimit - recalculation is needed.
_Recalculation_ : _amountUSDC = (500000 - (100000 * ONE_HBT_IN_WEI)) * 15 * ONE_HBT_IN_WEI / 100 =? - I think there will be a mistake.

**B2**: _amountUSDC = 75000, amountSHBT = 75000 * ONE_HBT_IN_WEI / (15 * ONE_HBT_IN_WEI / 100) = 500000. Since amountSHBT > purchaseLimit - recalculation is needed.
_Recalculation_ : amountSHBT = 100000, _amountUSDC = 100000 * (15 * ONE_HBT_IN_WEI / 100) / ONE_HBT_IN_WEI = 15000. The buyer is debited not 75,000 USDC, but 15,000 and he will buy 100,000 SHBT == purchaseLimit